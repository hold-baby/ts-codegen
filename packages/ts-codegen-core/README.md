# TS Codegen Core

[![Build Status](https://img.shields.io/travis/reeli/ts-codegen.svg?style=flat-square&branch=master)](https://travis-ci.org/reeli/ts-codegen)
[![codecov](https://codecov.io/gh/reeli/ts-codegen/branch/master/graph/badge.svg?style=flat-square)](https://codecov.io/gh/reeli/ts-codegen)
[![License](https://img.shields.io/npm/l/@ts-tool/ts-codegen.svg?style=flat-square)](https://npmjs.org/package/@ts-tool/ts-codegen)

Generate ts code from swagger.

# 开始

1. `npm install @ts-codegen-core`
