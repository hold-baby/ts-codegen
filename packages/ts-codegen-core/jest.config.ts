export default {
  testEnvironment: "node",
  modulePaths: ["<rootDir>"],
  moduleNameMapper: {
    "^@ts-tool/([^/]+)$": "<rootDir>/src/index.ts",
  },
  coverageDirectory: "coverage",
  testRegex: ".*/__tests__/.+\\.(generator|test|spec)\\.(ts|tsx)$",
  collectCoverage: true,
  collectCoverageFrom: ["src/**/*.{ts,tsx}", "!src/__types__/**", "!src/bootstrap.ts", "!src/index.ts"],
  transform: {
    "^.+\\.tsx?$": "ts-jest",
  },
};
