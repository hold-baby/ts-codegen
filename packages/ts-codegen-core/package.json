{
  "name": "@ts-tool/ts-codegen-core",
  "version": "3.3.0",
  "main": "dist/lib/index.js",
  "module": "dist/module/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist",
    "README.md"
  ],
  "scripts": {
    "start": "ts-node -r tsconfig-paths/register src/bootstrap.ts"
  },
  "dependencies": {
    "axios": "0.21.x",
    "js-yaml": "4.0.x",
    "lodash": "4.17.x",
    "prettier": "*"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "reeli",
  "license": "ISC",
  "bugs": {
    "url": "https://github.com/reeli/ts-codegen/issues"
  },
  "homepage": "https://github.com/reeli/ts-codegen#readme"
}
