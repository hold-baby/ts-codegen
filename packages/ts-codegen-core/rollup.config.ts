import path from "path";
// @ts-ignore
import rollupBabel from "rollup-plugin-babel";
// @ts-ignore
import rollupTypeScript from "rollup-plugin-typescript";

const pkg = require(path.join(process.cwd(), "package.json"));

module.exports = {
  input: "src/index.ts",
  output: [
    {
      dir: "dist/lib",
      format: "cjs",
    },
    {
      dir: "dist/module",
      format: "es",
    },
  ],
  external: [
    // @ts-ignore
    ...Object.keys(process.binding("natives")),
    ...Object.keys(pkg.dependencies || {}),
    ...Object.keys(pkg.peerDependencies || {}),
  ],
  plugins: [
    rollupTypeScript({
      target: "es5",
      module: "es6",
    }),
    rollupBabel({
      plugins: ["babel-plugin-pure-calls-annotation"],
      exclude: "node_modules/**",
      extensions: [".js", ".jsx", ".ts", ".tsx"],
    }),
  ],
};
