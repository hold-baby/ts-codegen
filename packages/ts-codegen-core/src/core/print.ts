import { IClientConfig, RequestType, ScanOptions } from "../__types__/types";
import { IStore, DeclKinds } from "./createRegister";
import { prettifyCode, objToTypeStr } from "../utils/common";
import { sortBy, isEmpty, compact, keys, mapValues, isString } from "lodash";
import { CustomType } from "./Type";
import { DEFAULT_SERVICE_NAME_IN_HEADER } from "..";

export const printOutputs = (
  clientConfigs: IClientConfig[],
  decls: IStore["decls"],
  requestCreateMethod: string = "createRequest",
  options?: ScanOptions,
) => {
  return prettifyCode(`${printRequest(clientConfigs, requestCreateMethod, options)} \n\n ${printTypes(decls)}`);
};

const hasRequestBody = (TReq: IClientConfig["TReq"]) => TReq?.requestBody || (TReq || {})["requestBody?"];

const printRequest = (clientConfigs: IClientConfig[], requestCreateMethod: string, options?: ScanOptions): string => {
  const configs = sortBy(clientConfigs, (o) => o.operationId);

  return configs
    .map((v) => {
      const toUrl = () => `url: \`${v.url}\`,`;
      const toMethod = () => `method: "${v.method}",`;
      const toRequestBody = () => {
        if (!isEmpty(v.bodyParams)) {
          return `data: ${v.bodyParams!.length > 1 ? `{${v.bodyParams!.join(",")}}` : v.bodyParams![0]},`;
        }
        return v.contentType && hasRequestBody(v.TReq) ? "data: requestBody," : "";
      };
      const toQueryParams = () => {
        const params = toRequestParams(v.queryParams);
        return params ? `params: ${params},` : "";
      };
      const toHeaders = () => {
        if (v.contentType && options?.withServiceNameInHeader) {
          return `headers: { 'Content-Type': '${v.contentType}', '${
            isString(options?.withServiceNameInHeader)
              ? options?.withServiceNameInHeader
              : DEFAULT_SERVICE_NAME_IN_HEADER
          }': serviceName },`;
        }

        if (v.contentType) {
          return `headers: { 'Content-Type': '${v.contentType}' },`;
        }

        if (options?.withServiceNameInHeader) {
          return `headers: { '${
            isString(options?.withServiceNameInHeader)
              ? options?.withServiceNameInHeader
              : DEFAULT_SERVICE_NAME_IN_HEADER
          }': serviceName },`;
        }

        return "";
      };

      const toGenerators = () => {
        const TReq = generateTReq(v.TReq);
        const TResp = v.TResp?.toType(false);

        if (!TReq && !TResp) {
          return "";
        }
        if (!TResp) {
          return `<${TReq}>`;
        }
        return `<${TReq}, ${TResp}>`;
      };
      const toRequestInputs = () => {
        const getRequestBody = () => {
          if (!isEmpty(v.bodyParams)) {
            return v.bodyParams!;
          }
          return v.contentType && hasRequestBody(v.TReq) ? ["requestBody"] : "";
        };
        const list = compact([...v.pathParams, ...v.queryParams, ...getRequestBody()]);
        return isEmpty(list) ? "" : toRequestParams(list);
      };

      return `
${addComments(v, options?.withComments)}
export const ${v.operationId} = ${requestCreateMethod}${toGenerators()}("${
        v.operationId
      }", (${toRequestInputs()}) => ({${toUrl()}${toMethod()}${toRequestBody()}${toQueryParams()}${toHeaders()}})
);
`;
    })
    .join("");
};

const addComments = (v: IClientConfig, withComments?: boolean) => {
  const summaryComment = withComments && v.summary ? `* ${v.summary}` : "";
  const deprecatedComment = v.deprecated ? `* @deprecated ${v.operationId}` : "";
  const comments = [summaryComment, deprecatedComment].filter((c) => !!c);

  return isEmpty(comments)
    ? ""
    : `/**
  ${comments.join("\\n")}
  */`;
};

const printTypes = (decls: IStore["decls"]): string => {
  return keys(decls)
    .sort()
    .map((k) => {
      const expr = decls[k].kind === DeclKinds.type ? "=" : "";
      const semi = decls[k].kind === DeclKinds.type ? ";" : "";
      return `export ${decls[k].kind} ${decls[k].name} ${expr} ${decls[k].type.toType()}${semi}`;
    })
    .join("\n\n");
};

const generateTReq = (TReq: IClientConfig["TReq"]) => {
  if (isEmpty(TReq)) {
    return;
  }

  function gen(obj: IClientConfig["TReq"]): string {
    return objToTypeStr(
      mapValues(obj, (v) => {
        if (!v.toType) {
          return gen(v as RequestType);
        }
        return (v as CustomType).toType(false);
      }),
    );
  }

  return gen(TReq);
};

const toRequestParams = (data: string[]) => (!isEmpty(data) ? `{\n ${data.join(",\n")} \n}` : undefined);
