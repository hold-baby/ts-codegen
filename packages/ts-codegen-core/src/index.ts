export * from "./core";
export * from "./utils";
export * from "./constants";
export * from "./codegen";
export * from "./__types__/OpenAPI";
export * from "./__types__/types";
