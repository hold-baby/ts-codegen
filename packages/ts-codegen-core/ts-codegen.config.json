{
  "outputFolder": "clients",
  "requestCreateLib": "../examples/utils/createRequest",
  "requestCreateMethod": "createRequest",
  "apiSpecsPaths": [
    {
      "path": "https://petstore.swagger.io/v2/swagger.json",
      "name": "PetStoreService1"
    },
    {
      "path": "https://petstore.swagger.io/v2/swagger.yaml",
      "name": "PetStoreService2"
    },
    {
      "path": "./examples/swagger.json",
      "name": "SwaggerService"
    },
    {
      "path": "./examples/openapi.json",
      "name": "OpenApiService"
    },
    {
      "path": "./examples/openapi1.json",
      "name": "OpenApiService1"
    },
    {
      "path": "./examples/demo.yaml",
      "name": "DemoService"
    }
  ],
  "options": {
    "withComments": true,
    "typeWithPrefix": true,
    "backwardCompatible": true,
    "withServiceNameInHeader": true
  }
}
