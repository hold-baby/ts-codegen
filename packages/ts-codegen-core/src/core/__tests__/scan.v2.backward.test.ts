import swaggerV2 from "../../../examples/swagger.json";
import { Spec } from "swagger-schema-official";
import { scan } from "@ts-tool/ts-codegen-core";
import { printOutputs } from "../print";

describe("swagger v2", () => {
  it("should handle spec without prefix in type name", () => {
    const { clientConfigs, decls } = scan(swaggerV2 as Spec, { typeWithPrefix: true, backwardCompatible: true });
    expect(printOutputs(clientConfigs, decls)).toMatchSnapshot();
  });
});
