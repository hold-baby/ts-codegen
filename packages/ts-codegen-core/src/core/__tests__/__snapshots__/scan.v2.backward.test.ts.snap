// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swagger v2 should handle spec without prefix in type name 1`] = `
"export const addPet = createRequest<{
  body: IPet;
}>(\\"addPet\\", ({ body }) => ({
  url: \`/api/v2/pet\`,
  method: \\"post\\",
  data: body,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const createItem = createRequest<
  {
    item?: IItem;
  },
  IMessage
>(\\"createItem\\", ({ item }) => ({
  url: \`/api/v2/items\`,
  method: \\"post\\",
  data: item,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const createUser = createRequest<{
  body: IUser;
}>(\\"createUser\\", ({ body }) => ({
  url: \`/api/v2/user\`,
  method: \\"post\\",
  data: body,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const createUsersWithArrayInput = createRequest<{
  body: IUser[];
}>(\\"createUsersWithArrayInput\\", ({ body }) => ({
  url: \`/api/v2/user/createWithArray\`,
  method: \\"post\\",
  data: body,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const createUsersWithListInput = createRequest<{
  body: IUser[];
}>(\\"createUsersWithListInput\\", ({ body }) => ({
  url: \`/api/v2/user/createWithList\`,
  method: \\"post\\",
  data: body,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const deleteAttachmentUsingDELETE = createRequest<{
  id: string;
}>(\\"deleteAttachmentUsingDELETE\\", ({ id }) => ({ url: \`/api/v2/\${id}\`, method: \\"delete\\" }));

export const deleteAttachmentUsingDELETE2 = createRequest<{
  id: string;
}>(\\"deleteAttachmentUsingDELETE2\\", ({ id }) => ({ url: \`/api/v2/test/\${id}\`, method: \\"delete\\" }));

export const deleteOrder = createRequest<{
  orderId: number;
}>(\\"deleteOrder\\", ({ orderId }) => ({ url: \`/api/v2/store/order/\${orderId}\`, method: \\"delete\\" }));

export const deletePet = createRequest<{
  petId: number;
}>(\\"deletePet\\", ({ petId }) => ({ url: \`/api/v2/pet/\${petId}\`, method: \\"delete\\" }));

export const deleteUser = createRequest<{
  username: string;
}>(\\"deleteUser\\", ({ username }) => ({ url: \`/api/v2/user/\${username}\`, method: \\"delete\\" }));

export const downloadUsingGET = createRequest<
  {
    id: string;
  },
  IResource
>(\\"downloadUsingGET\\", ({ id }) => ({ url: \`/api/v2/\${id}\`, method: \\"get\\" }));

export const downloadUsingGET2 = createRequest<
  {
    id: string;
    idInPath: string;
  },
  IResource
>(\\"downloadUsingGET2\\", ({ idInPath, id }) => ({
  url: \`/api/v2/test/\${idInPath}\`,
  method: \\"get\\",
  params: {
    id,
  },
}));

export const findBookByIdUsingGET = createRequest<
  {
    id: string;
  },
  IBookDetailVo
>(\\"findBookByIdUsingGET\\", ({ id }) => ({ url: \`/api/v2/book/\${id}\`, method: \\"get\\" }));

export const findPetsByStatus = createRequest<
  {
    status: (keyof typeof FindPetsByStatusStatus)[];
  },
  IPet[]
>(\\"findPetsByStatus\\", ({ status }) => ({
  url: \`/api/v2/pet/findByStatus\`,
  method: \\"get\\",
  params: {
    status,
  },
}));

/**
 * @deprecated findPetsByTags
 */
export const findPetsByTags = createRequest<
  {
    tags: string[];
  },
  IPet[]
>(\\"findPetsByTags\\", ({ tags }) => ({
  url: \`/api/v2/pet/findByTags\`,
  method: \\"get\\",
  params: {
    tags,
  },
}));

export const getAgencyResponseListUsingGET = createRequest<
  {
    addressType: keyof typeof GetAgencyResponseListUsingGetAddressType;
  },
  IResource[]
>(\\"getAgencyResponseListUsingGET\\", ({ addressType }) => ({
  url: \`/api/v2/agency-responses\`,
  method: \\"get\\",
  params: {
    addressType,
  },
}));

export const getAgencyResponseUsingGET = createRequest<
  {
    addressType: keyof typeof GetAgencyResponseUsingGetAddressType;
    carType: CarTypeEnum;
    responseDate: string;
  },
  IResource
>(\\"getAgencyResponseUsingGET\\", ({ responseDate, addressType, carType }) => ({
  url: \`/api/v2/agency-response\`,
  method: \\"get\\",
  data: carType,
  params: {
    responseDate,
    addressType,
  },
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const getDocumentByIdUsingGET = createRequest<
  {
    documentId: string;
    from?: keyof typeof GetDocumentByIdUsingGetFrom;
  },
  IDocumentVo
>(\\"getDocumentByIdUsingGET\\", ({ documentId, from }) => ({
  url: \`/api/v2/documents/\${documentId}/doc\`,
  method: \\"get\\",
  params: {
    from,
  },
}));

export const getHealthCheckUsingHEAD = createRequest<
  {
    flight_no: string;
  },
  IDocumentVo
>(\\"getHealthCheckUsingHEAD\\", ({ flight_no }) => ({ url: \`/api/v2/health-check/\${flight_no}\`, method: \\"head\\" }));

export const getInventory = createRequest<undefined, { [key: string]: number }>(\\"getInventory\\", () => ({
  url: \`/api/v2/store/inventory\`,
  method: \\"get\\",
}));

export const getItems = createRequest<
  {
    limit?: number;
    page?: number;
    sort?: string[];
  },
  TItems
>(\\"getItems\\", ({ page, limit, sort }) => ({
  url: \`/api/v2/items\`,
  method: \\"get\\",
  params: {
    page,
    limit,
    sort,
  },
}));

export const getOrderById = createRequest<
  {
    orderId: number;
  },
  IOrder
>(\\"getOrderById\\", ({ orderId }) => ({ url: \`/api/v2/store/order/\${orderId}\`, method: \\"get\\" }));

export const getPetById = createRequest<
  {
    petId: number;
  },
  IPet
>(\\"getPetById\\", ({ petId }) => ({ url: \`/api/v2/pet/\${petId}\`, method: \\"get\\" }));

export const getScheduleDetailsByDateUsingGET = createRequest<
  {
    roleId?: string;
    scheduleDate: number;
  },
  IScheduleVo[]
>(\\"getScheduleDetailsByDateUsingGET\\", ({ scheduleDate, roleId }) => ({
  url: \`/api/v2/schedules\`,
  method: \\"get\\",
  params: {
    scheduleDate,
    roleId,
  },
}));

export const getTeamListByUsingGET = createRequest<{
  limit?: number;
  offset?: number;
}>(\\"getTeamListByUsingGET\\", ({ offset, limit }) => ({
  url: \`/api/v2/teams\`,
  method: \\"get\\",
  params: {
    offset,
    limit,
  },
}));

export const getUserByName = createRequest<
  {
    username: string;
  },
  IUser
>(\\"getUserByName\\", ({ username }) => ({ url: \`/api/v2/user/\${username}\`, method: \\"get\\" }));

export const getUserListByUsingGET = createRequest<{
  limit?: number;
  offset?: number;
}>(\\"getUserListByUsingGET\\", ({ offset, limit }) => ({
  url: \`/api/v2/users\`,
  method: \\"get\\",
  params: {
    offset,
    limit,
  },
}));

export const loginUser = createRequest<
  {
    password: string;
    username: string;
  },
  string
>(\\"loginUser\\", ({ username, password }) => ({
  url: \`/api/v2/user/login\`,
  method: \\"get\\",
  params: {
    username,
    password,
  },
}));

export const logoutUser = createRequest(\\"logoutUser\\", () => ({ url: \`/api/v2/user/logout\`, method: \\"get\\" }));

export const placeOrder = createRequest<
  {
    body: IOrder;
  },
  IOrder
>(\\"placeOrder\\", ({ body }) => ({
  url: \`/api/v2/store/order\`,
  method: \\"post\\",
  data: body,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const updateBookByIdUsingPUT = createRequest<{
  id: string;
  updateBookRequest: IUpdateBookRequest;
}>(\\"updateBookByIdUsingPUT\\", ({ id, updateBookRequest }) => ({
  url: \`/api/v2/book/\${id}\`,
  method: \\"put\\",
  data: updateBookRequest,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const updatePet = createRequest<{
  body: IPet;
}>(\\"updatePet\\", ({ body }) => ({
  url: \`/api/v2/pet\`,
  method: \\"put\\",
  data: body,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const updatePetWithForm = createRequest<{
  name?: string;
  petId: number;
  status?: string;
}>(\\"updatePetWithForm\\", ({ petId, name, status }) => ({
  url: \`/api/v2/pet/\${petId}\`,
  method: \\"post\\",
  data: { name, status },
  headers: { \\"Content-Type\\": \\"application/x-www-form-urlencoded\\" },
}));

export const updateUser = createRequest<{
  body: IUser;
  username: string;
}>(\\"updateUser\\", ({ username, body }) => ({
  url: \`/api/v2/user/\${username}\`,
  method: \\"put\\",
  data: body,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const uploadAttachmentUsingPOST = createRequest<
  {
    attachment: Blob;
  },
  IAttachmentBo
>(\\"uploadAttachmentUsingPOST\\", ({ attachment }) => ({
  url: \`/api/v2\`,
  method: \\"post\\",
  data: attachment,
  headers: { \\"Content-Type\\": \\"multipart/form-data\\" },
}));

export const uploadFile = createRequest<
  {
    additionalMetadata?: string;
    file?: Blob;
    petId: number;
  },
  IApiResponse
>(\\"uploadFile\\", ({ petId, additionalMetadata, file }) => ({
  url: \`/api/v2/pet/\${petId}/uploadImage\`,
  method: \\"post\\",
  data: { additionalMetadata, file },
  headers: { \\"Content-Type\\": \\"multipart/form-data\\" },
}));

export interface IApiResponse {
  code?: number;
  message?: string;
  type?: string;
}

export interface IAttachmentBo {
  authorName?: string;
  createdDate?: number;
  fileName?: string;
  id?: string;
  mimeType?: string;
  path?: string;
}

export interface IBookDetailVo {
  attachment?: IScheduleVo;
  authorName?: string;
  createdDate?: number;
  fileName?: string;
  id?: string;
  mimeType?: string;
  path?: string;
}

export interface IBookVo {
  address?: string;
  price?: string;
}

export interface IBookingResponse {
  data: IDocumentVo;
  errors?: IErrorInfo[];
}

export enum CarTypeEnum {
  \\"BMW\\" = \\"BMW\\",
  \\"Audi\\" = \\"Audi\\",
  \\"BYD\\" = \\"BYD\\",
}

export interface ICategory {
  id?: number;
  name?: string;
}

export interface IDictionary {
  category?: ICategory;
  dictionary1?: { age: number; name: string; [key: string]: string | number };
  dictionary2?: { age: string; name: string; [key: string]: string };
  dictionary3?: { age?: number; name?: string; [key: string]: any };
  dictionary4?: { [key: string]: any };
  dictionary?: { [key: string]: { code?: number; text?: string } };
  id?: number;
  name: string;
  photoUrls: string[];
  status?: keyof typeof DictionaryStatus;
  tags?: ITag[];
}

export enum DictionaryStatus {
  \\"available\\" = \\"available\\",
  \\"pending\\" = \\"pending\\",
  \\"sold\\" = \\"sold\\",
}

export interface IDocumentVo {
  attachment?: IBookDetailVo;
  authorName?: string;
  createdDate?: number;
  id?: string;
  note?: string;
  title?: string;
}

export interface IError {
  code?: number;
  field?: string;
  message?: string;
}

export interface IErrorInfo {
  errorMessage?: string;
}

export type TErrors = IError[];

export interface IFile {
  absolute?: boolean;
  absoluteFile?: IFile;
  absolutePath?: string;
  canonicalFile?: IFile;
  canonicalPath?: string;
  directory?: boolean;
  executable?: boolean;
  file?: boolean;
  freeSpace?: number;
  hidden?: boolean;
  lastModified?: number;
  name?: string;
  parent?: string;
  parentFile?: IFile;
  path?: string;
  readable?: boolean;
  totalSpace?: number;
  usableSpace?: number;
  writable?: boolean;
}

export enum FindPetsByStatusStatus {
  \\"available\\" = \\"available\\",
  \\"pending\\" = \\"pending\\",
  \\"sold\\" = \\"sold\\",
}

export enum GetAgencyResponseListUsingGetAddressType {
  \\"AAA\\" = \\"AAA\\",
  \\"BBB\\" = \\"BBB\\",
}

export enum GetAgencyResponseUsingGetAddressType {
  \\"AAA\\" = \\"AAA\\",
  \\"BBB\\" = \\"BBB\\",
}

export enum GetDocumentByIdUsingGetFrom {
  \\"AAA\\" = \\"AAA\\",
  \\"BBB\\" = \\"BBB\\",
}

export interface IInputStream {
  [key: string]: any;
}

export interface IItem {
  color?: string;
  id?: string;
  name?: string;
  price?: string;
}

export type TItems = IItem[];

export interface IMessage {
  message?: string;
}

export interface IOrder {
  complete?: boolean;
  id?: number;
  petId?: number;
  quantity?: number;
  shipDate?: string;
  status?: keyof typeof OrderStatus;
}

export enum OrderStatus {
  \\"placed\\" = \\"placed\\",
  \\"approved\\" = \\"approved\\",
  \\"delivered\\" = \\"delivered\\",
}

export interface IPet {
  category?: ICategory;
  id?: number;
  name: string;
  photoUrls: string[];
  status?: keyof typeof PetStatus;
  tags?: ITag[];
}

export enum PetStatus {
  \\"available\\" = \\"available\\",
  \\"pending\\" = \\"pending\\",
  \\"sold\\" = \\"sold\\",
}

export interface IResource {
  description?: string;
  file?: IFile;
  filename?: string;
  inputStream?: IInputStream;
  open?: boolean;
  readable?: boolean;
  uri?: IUri;
  url?: IUrl;
}

export interface IScheduleVo {
  contentType?: TScheduleVoContentType;
  reason?: TScheduleVoReason;
  schedules?: IBookVo[][];
  shiftId?: string;
  team?: string;
}

export type TScheduleVoContentType = \\"01\\" | \\"02\\" | \\"03\\" | \\"04\\" | \\"05\\" | \\"06\\";

export type TScheduleVoReason = \\"1\\" | \\"3\\" | \\"4\\" | \\"5\\" | \\"6\\";

export interface ITag {
  id?: number;
  name?: string;
}

export interface IUpdateBookRequest {
  birthCountry?: string;
  citizenship?: string;
  dateOfBirth?: number;
  employmentStatus?: string;
  ethnicity?: string;
  gender?: string;
  idNumber?: string;
  idType?: string;
  roleId?: string;
  spokenLanguage?: string[];
}

export interface IUri {
  absolute?: boolean;
  authority?: string;
  fragment?: string;
  host?: string;
  opaque?: boolean;
  path?: string;
  port?: number;
  query?: string;
  rawAuthority?: string;
  rawFragment?: string;
  rawPath?: string;
  rawQuery?: string;
  rawSchemeSpecificPart?: string;
  rawUserInfo?: string;
  scheme?: string;
  schemeSpecificPart?: string;
  userInfo?: string;
}

export interface IUrl {
  authority?: string;
  content?: { [key: string]: any };
  defaultPort?: number;
  deserializedFields?: IUrlStreamHandler;
  file?: string;
  host?: string;
  path?: string;
  port?: number;
  protocol?: string;
  query?: string;
  ref?: string;
  serializedHashCode?: number;
  userInfo?: string;
}

export interface IUrlStreamHandler {
  [key: string]: any;
}

export interface IUser {
  email?: string;
  firstName?: string;
  id?: number;
  lastName?: string;
  password?: string;
  phone?: string;
  userStatus?: number;
  username?: string;
}

export interface IValidationError {
  errors?: TErrors;
  message?: string;
}
"
`;
